pgjdbc/src/main/java/org/postgresql/jdbcurlresolver/JdbcUrlResolver.java

```diff
@@ -0,0 +1,326 @@
+ */
+
+package org.postgresql.jdbcurlresolver;
+
+import static org.postgresql.jdbcurlresolver.JdbcUrlParser.saveProperty;
+import static org.postgresql.util.internal.Nullness.castNonNull;
+
+import org.postgresql.PGEnvironment;
+import org.postgresql.PGProperty;
+import org.postgresql.util.OSUtil;
+
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Map;
+import java.util.Properties;
+import java.util.TreeMap;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import java.util.stream.Collectors;
+
+/**
+ * parser for jdbc URL.
+ * https://www.postgresql.org/docs/current/libpq-connect.html#LIBPQ-CONNSTRING
+ */
+public class JdbcUrlResolver {
+  private static final Logger LOGGER = Logger.getLogger(JdbcUrlResolver.class.getName());
+
+  // input
+  private final String url;
+  private final Properties defaults;
+
+  // to support unit tests
+  private @Nullable JdbcUrlResolverFatalException failException;
+
+  // to support unit tests
+  @Nullable JdbcUrlResolverFatalException getFailException() {
+    return failException;
+  }
+
+  /**
+   * @param url      jdbc url
+   * @param defaults defaults
+   */
+  public JdbcUrlResolver(String url, @Nullable Properties defaults) {
+    this.url = url;
+    this.defaults = defaults == null ? new Properties() : defaults;
+  }
+
+  /**
+   * Returns Properties
+   *
+   * @return result of parsing URL
+   */
+  public @Nullable Properties getResult() {
+    Properties result = new Properties();
+    try {
+      // override hierarchy
+      Properties p8 = new Properties(); // global defaults
+      Properties p7 = new Properties(p8); // org/postgresql/driverconfig.properties by classloader
+      Properties p6 = new Properties(p7); // os env
+      Properties p5 = new Properties(p6); // java env
+      Properties p4 = new Properties(p5); // service
+      Properties p3 = new Properties(p4); // Properties given to getConnection()
+      Properties p2 = new Properties(p3); // URL values
+      Properties p1 = new Properties(p2); // URL arguments
+      Properties p0 = new Properties(p1); // Calculated overrides
+      // priority 1: URL arguments
+      // priority 2: URL values
+      JdbcUrlParser parser = new JdbcUrlParser(url);
+      parser.parse(p1, p2);
+      // priority 3: Properties given as argument to DriverManager.getConnection()
+      // argument "defaults" EXCLUDING defaults
+      parse3GetConnectionProperties(p3);
+      // priority 5: Java PGEnvironment values
+      parse5JavaEnvironment(p5);
+      // priority 6: OS PGEnvironment values
+      parse6OsEnvironment(p6);
+      // priority 7: default values of Properties given as argument to DriverManager.getConnection()
+      //             source: org/postgresql/driverconfig.properties by classloader
+      // argument "defaults" ONLY defaults
+      parse7DraiverconfigProperties(p7);
```

Comment:
DraiverconfigPropertis is probably misspelled ?

Reply:
will fix